staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -70,24 +108,40 @@ func (b *backend) logEvent(ev *auditinternal.Event) {
 	case FormatJson:
 		bs, err := runtime.Encode(audit.Codecs.LegacyCodec(b.groupVersion), ev)
 		if err != nil {
-			audit.HandlePluginError("log", err, ev)
+			audit.HandlePluginError(pluginName, err, ev)
 			return
 		}
 		line = string(bs[:])
 	default:
-		audit.HandlePluginError("log", fmt.Errorf("log format %q is not in list of known formats (%s)",
+		audit.HandlePluginError(pluginName, fmt.Errorf("log format %q is not in list of known formats (%s)",
 			b.format, strings.Join(AllowedFormats, ",")), ev)
 		return
 	}
 	if _, err := fmt.Fprint(b.out, line); err != nil {
-		audit.HandlePluginError("log", err, ev)
+		audit.HandlePluginError(pluginName, err, ev)
+	}
+}
+
+func (b *backend) worker() {
+	for {
+		event, ok := <-b.buffer
+		if !ok {
+			return
+		}
+		b.logEvent(event)
 	}
 }
 
 func (b *backend) Run(stopCh <-chan struct{}) error {
+	glog.Infof("Starting audit log backend")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(b.worker, time.Second, stopCh)
+	}
+
 	return nil
 }
 
 func (b *backend) Shutdown() {
-	// Nothing to do here.
+	close(b.buffer)
```

Comment:
It might be good to block here until the channel is drained. You could use a waitgroup to check when the worker is done.

Reply:
close channel here is a signal for worker when the buffer drained and closed. They will all exit. 
If some event still in channel, then close channel, the workers can still receive the event and write it to file, when buffer drained, the workers will exit.

As you said block here until buffer is drained, I think need add a few notification mechanism.

Reply:
As I mentioned, you can just use a waitgroup for the notification mechanism. The reason you want to block here is that you don't want to let the process exit before the worker has had a chance to drain the buffer. If the process exits, all goroutines will be killed.